### PR TITLE
refactor!: Split `IssueRequest` into `CreateIssueRequest` & `UpdateIssueRequest` and pass by value

### PR DESCRIPTION
**Pull request by jvm986**
BREAKING CHANGE: `IssueService.Edit` is renamed to `IssueService.Update` to match the underlying "update an issue" endpoint.

`IssuesService.Create` now takes a dedicated `CreateIssueRequest` by value (with a required non-pointer `Title`).
`IssuesService.Updated` takes `UpdateIssueRequest` by value.

Towards #3644.

### CODE REVIEW

**Review by gmlewis — APPROVED** [review]
Thank you, @jvm986!
LGTM.
Awaiting second LGTM+Approval from any other contributor to this repo before merging.

cc: @stevehipwell - @alexandear - @Not-Dhananjay-Mishra - @JamBalaya56562 

Oh, and @jvm986 and @JamBalaya56562 - feel free to "cc:" each other on each PR to be extra careful that work is not duplicated.

**Comment by jvm986** [comment]
@gmlewis, @JamBalaya56562 

What are your thoughts on the `CreateIssueRequest` question?

> IssueRequest is shared by Create (API requires title) and Edit (all optional), so I only changed the param to a value — promoting Title to a required string would break Edit. Should we split into a dedicated CreateIssueRequest (like https://github.com/google/go-github/pull/4382 did for deployment branch policies), or is the shared type fine here?

**Comment by gmlewis** [comment]
> @gmlewis, @JamBalaya56562
> 
> What are your thoughts on the `CreateIssueRequest` question?
> 
> > IssueRequest is shared by Create (API requires title) and Edit (all optional), so I only changed the param to a value — promoting Title to a required string would break Edit. Should we split into a dedicated CreateIssueRequest (like #4382 did for deployment branch policies), or is the shared type fine here?

I think there was a discussion for #4382 where the change of name was deemed valuable, so I think we should also address that here in the breaking change PR to remain consistent.

Thank you for bringing this up!  I appreciate it.

File: github/issues.go [hunk 1]
```diff
@@ -104,7 +104,22 @@ func (i Issue) IsPullRequest() bool {
...
+	Milestone        *int                      `json:"milestone,omitempty"`
+	Assignees        []string                  `json:"assignees,omitempty"`
+	Type             *string                   `json:"type,omitempty"`
+	IssueFieldValues []*IssueRequestFieldValue `json:"issue_field_values,omitempty"`
+}
+
+// IssueRequest represents a request to edit an issue.
 // It is separate from Issue above because otherwise Labels
 // and Assignee fail to serialize to the correct JSON.
 type IssueRequest struct {
```
Comment (Not-Dhananjay-Mishra) [inline github/issues.go]: Can we update `IssueRequest` struct too. 
It has one missing field `duplicate_issue_id` and we should use `[]string` instead of `*[]string` for `Labels` and `Assignees`

File: github/issues.go [hunk 2]
```diff
@@ -479,7 +494,7 @@ func (s *IssuesService) Create(ctx context.Context, owner, repo string, body *Is
 // GitHub API docs: https://docs.github.com/rest/issues/issues?apiVersion=2022-11-28#update-an-issue
 //
 //meta:operation PATCH /repos/{owner}/{repo}/issues/{issue_number}
-func (s *IssuesService) Edit(ctx context.Context, owner, repo string, number int, body *IssueRequest) (*Issue, *Response, error) {
+func (s *IssuesService) Edit(ctx context.Context, owner, repo string, number int, body IssueRequest) (*Issue, *Response, error) {
```
Comment (Not-Dhananjay-Mishra) [inline github/issues.go]: Can we rename this to `Update`? as API docs mentioned it as ["Update an issue"](https://docs.github.com/en/rest/issues/issues?apiVersion=2022-11-28#update-an-issue--code-samples)

File: github/issues.go (hunk 1 above)
Comment (jvm986) [reply github/issues.go]: Nice find on `duplicate_issue_id`, will add it.

Regarding `Labels` and `Assignees` the logic here is that the update endpoint treats `"labels": []` as "clear all labels" (distinct from omitting the field which leaves them unchanged). `[]string` isn't able to distinguish between omission and an explicit empty array.

File: github/issues.go (hunk 2 above)
Comment (jvm986) [reply github/issues.go]: Makes sense to me, will update after feedback from others.

Comment (gmlewis) [reply github/issues.go]: > Makes sense to me, will update after feedback from others.

SGTM. Thank you, @Not-Dhananjay-Mishra and @jvm986!

I figure that while we are actually breaking the API, we should always attempt to make it as easy-to-understand and easy-to-use as possible, and this seems to fit into that category. :smile:

File: github/issues.go (hunk 1 above)
Comment (Not-Dhananjay-Mishra) [reply github/issues.go]: @gmlewis could you help me with this? I assumed that we don't use `*[]buildin` in this repo. 

> Regarding `Labels` and `Assignees` the logic here is that the update endpoint treats `"labels": []` as "clear all labels" (distinct from omitting the field which leaves them unchanged). `[]string` isn't able to distinguish between and an explicit empty array.

`omitzero` seems like a good fit here instead of `omitempty`, since it omits only `nil` slices while preserving an explicit empty slice.

Comment (gmlewis) [reply github/issues.go]: > `omitzero` seems like a good fit here instead of `omitempty`, since it omits only `nil` slices while preserving an explicit empty slice.

Yes, exactly right, @Not-Dhananjay-Mishra - this is what `omitzero` is for. So we can finally get rid of the `*[]string`, change it to `[]string` and then use `omitzero` on it so that `nil` is different from `[]string{}`.

Comment (Not-Dhananjay-Mishra) [reply github/issues.go]: Thanks @gmlewis.
@jvm986, you can change `*[]string` to `[]string` for `Labels` and `Assignees`, and use `omitzero` instead of `omitempty`.

Comment (gmlewis) [reply github/issues.go]: > Thanks @gmlewis. @jvm986, you can change `*[]string` to `[]string` for `Labels` and `Assignees`, and use `omitzero` instead of `omitempty`.

Yes, please, and just be absolutely sure to include a table-driven unit test where _BOTH_ `nil` and `[]string{}` are passed to ensure that the generated JSON matches our expections, and preferrably add a comment that explains when to use one versus the other. Thank you, @jvm986!

File: github/issues.go [hunk 3]
```diff
@@ -104,19 +104,40 @@ func (i Issue) IsPullRequest() bool {
...
-type IssueRequest struct {
-	Title    *string   `json:"title,omitempty"`
-	Body     *string   `json:"body,omitempty"`
-	Labels   *[]string `json:"labels,omitempty"`
-	Assignee *string   `json:"assignee,omitempty"`
-	State    *string   `json:"state,omitempty"`
-	// StateReason can be 'completed' or 'not_planned'.
-	StateReason      *string                   `json:"state_reason,omitempty"`
+type CreateIssueRequest struct {
+	// Title is required when creating an issue.
```
Comment (alexandear) [inline github/issues.go]: This comment can be removed.

File: github/issues.go [hunk 4]
```diff
@@ -474,12 +495,12 @@ func (s *IssuesService) Create(ctx context.Context, owner, repo string, body *Is
...
 }
 
-// Edit (update) an issue.
+// Update an issue.
 //
 // GitHub API docs: https://docs.github.com/rest/issues/issues?apiVersion=2022-11-28#update-an-issue
 //
 //meta:operation PATCH /repos/{owner}/{repo}/issues/{issue_number}
-func (s *IssuesService) Edit(ctx context.Context, owner, repo string, number int, body *IssueRequest) (*Issue, *Response, error) {
+func (s *IssuesService) Update(ctx context.Context, owner, repo string, number int, body IssueRequest) (*Issue, *Response, error) {
```
Comment (alexandear) [inline github/issues.go]: Let's rename this as well:
```suggestion
func (s *IssuesService) Update(ctx context.Context, owner, repo string, number int, body UpdateIssueRequest) (*Issue, *Response, error) {
```

Comment (jvm986) [reply github/issues.go]: Good call, as it's a breaking change anyway.

**Review by Not-Dhananjay-Mishra — APPROVED** [review]
LGTM 🚀
Just update Title and Description.

**Comment by gmlewis** [comment]
Thank you, @jvm986, @alexandear, and @Not-Dhananjay-Mishra!
Merging.
